File: lib/si-frontend-types-rs/src/func.rs
use serde::{Deserialize, Serialize};
use si_events::{
    ActionKind, ActionPrototypeId, AttributePrototypeArgumentId, AttributePrototypeId, ComponentId,
    FuncArgumentId, FuncId, FuncKind, InputSocketId, OutputSocketId, PropId, SchemaVariantId,
};
use strum::{AsRefStr, Display, EnumIter, EnumString};

#[derive(Clone, Debug, Deserialize, Eq, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct FuncSummary {
    pub func_id: FuncId,
    pub kind: FuncKind,
    pub name: String,
    pub display_name: Option<String>,
    pub is_locked: bool,
}
#[derive(Clone, Debug, Deserialize, Eq, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct FuncCode {
    pub func_id: FuncId,
    pub code: String,
    pub types: String,
}
#[derive(Clone, Debug, Deserialize, Eq, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct FuncBindings {
    pub bindings: Vec<FuncBinding>,
}
#[remain::sorted]
#[derive(
    AsRefStr, Clone, Debug, Deserialize, EnumString, Eq, Serialize, Display, EnumIter, PartialEq,
)]
#[serde(rename_all = "camelCase", tag = "bindingKind")]
pub enum FuncBinding {
    #[serde(rename_all = "camelCase")]
    Action {
        // unique ids
        schema_variant_id: Option<SchemaVariantId>,
        action_prototype_id: Option<ActionPrototypeId>,
        func_id: Option<FuncId>,
        //thing that can be updated
        kind: Option<ActionKind>,
    },
    #[serde(rename_all = "camelCase")]
    Attribute {
        // unique ids
        func_id: Option<FuncId>,
        attribute_prototype_id: Option<AttributePrototypeId>,
        // things needed for create
        component_id: Option<ComponentId>,
        schema_variant_id: Option<SchemaVariantId>,

        // things that can be updated
        prop_id: Option<PropId>,
        output_socket_id: Option<OutputSocketId>,
        argument_bindings: Vec<AttributeArgumentBinding>,
    },
    #[serde(rename_all = "camelCase")]
    Authentication {
        // unique ids
        schema_variant_id: SchemaVariantId,
        func_id: FuncId,
    },
    #[serde(rename_all = "camelCase")]
    CodeGeneration {
        // unique ids
        schema_variant_id: Option<SchemaVariantId>,
        component_id: Option<ComponentId>,
        func_id: Option<FuncId>,
        attribute_prototype_id: Option<AttributePrototypeId>,

        // thing that can be updated
        inputs: Vec<LeafInputLocation>,
    },
    #[serde(rename_all = "camelCase")]
    Qualification {
        // unique ids
        schema_variant_id: Option<SchemaVariantId>,
        component_id: Option<ComponentId>,
        func_id: Option<FuncId>,
        attribute_prototype_id: Option<AttributePrototypeId>,

        // thing that can be updated
        inputs: Vec<LeafInputLocation>,
    },
}

#[derive(Clone, Debug, Deserialize, Eq, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AttributeArgumentBinding {
    pub func_argument_id: FuncArgumentId,
    pub attribute_prototype_argument_id: Option<AttributePrototypeArgumentId>,
    pub prop_id: Option<PropId>,
    pub input_socket_id: Option<InputSocketId>,
}

/// This enum provides available child [`Prop`](crate::Prop) trees of [`RootProp`](crate::RootProp)
/// that can be used as "inputs" for [`Funcs`](crate::Func) on leaves.
///
/// _Note: not all [`children`](crate::RootPropChild) of [`RootProp`](crate::RootProp) can be used
/// as "inputs" in order to prevent cycles. This enum provides an approved subset of those
/// children_.
#[remain::sorted]
#[derive(
    AsRefStr,
    Clone,
    Copy,
    Debug,
    Deserialize,
    EnumString,
    Eq,
    Serialize,
    Display,
    EnumIter,
    Ord,
    PartialEq,
    PartialOrd,
)]
#[serde(rename_all = "camelCase")]
pub enum LeafInputLocation {
    /// The input location corresponding to "/root/code".
    Code,
    /// The input location corresponding to "/root/deleted_at"
    DeletedAt,
    /// The input location corresponding to "/root/domain".
    Domain,
    /// The input location corresponding to "/root/resource".
    Resource,
    /// The input location corresponding to "/root/secrets".
    Secrets,
}

File: app/web/src/api/sdf/dal/func.ts
import { ActionKind, ActionPrototypeId } from "@/api/sdf/dal/action";
import { OutputSocketId, SchemaVariantId } from "./schema";
import { ComponentId } from "./component";
import { PropId } from "./prop";

export type FuncArgumentId = string;
export type FuncId = string;
export type AttributePrototypeArgumentId = string;

export enum FuncKind {
  Action = "Action",
  Attribute = "Attribute",
  Authentication = "Authentication",
  CodeGeneration = "CodeGeneration",
  Intrinsic = "Intrinsic",
  Qualification = "Qualification",
  SchemaVariantDefinition = "SchemaVariantDefinition",
  Unknown = "Unknown",
}

export enum CustomizableFuncKind {
  Action = "Action",
  Attribute = "Attribute",
  Authentication = "Authentication",
  CodeGeneration = "CodeGeneration",
  Qualification = "Qualification",
}

// TODO(nick,wendy): this is ugly to use in some places. We probably need to think of a better interface. Blame me, not Wendy.
export function customizableFuncKindToFuncKind(
  customizableFuncKind: CustomizableFuncKind,
): FuncKind {
  switch (customizableFuncKind) {
    case CustomizableFuncKind.Action:
      return FuncKind.Action;
    case CustomizableFuncKind.Attribute:
      return FuncKind.Attribute;
    case CustomizableFuncKind.Authentication:
      return FuncKind.Authentication;
    case CustomizableFuncKind.CodeGeneration:
      return FuncKind.CodeGeneration;
    case CustomizableFuncKind.Qualification:
      return FuncKind.Qualification;
    default:
      throw new Error(
        "this should not be possible since CustomizableFuncKind is a subset of FuncKind",
      );
  }
}

export const CUSTOMIZABLE_FUNC_TYPES = {
  [CustomizableFuncKind.Action]: {
    pluralLabel: "Actions",
    singularLabel: "Action",
  },
  [CustomizableFuncKind.Attribute]: {
    pluralLabel: "Attributes",
    singularLabel: "Attribute",
  },
  [CustomizableFuncKind.Authentication]: {
    pluralLabel: "Authentications",
    singularLabel: "Authentication",
  },
  [CustomizableFuncKind.CodeGeneration]: {
    pluralLabel: "Code Generations",
    singularLabel: "Code Generation",
  },
  [CustomizableFuncKind.Qualification]: {
    pluralLabel: "Qualifications",
    singularLabel: "Qualification",
  },
};

export const isCustomizableFuncKind = (f: FuncKind) =>
  f in CUSTOMIZABLE_FUNC_TYPES;

export enum FuncArgumentKind {
  Array = "Array",
  Boolean = "Boolean",
  Integer = "Integer",
  Json = "Json",
  Object = "Object",
  String = "String",
  Map = "Map",
  Any = "Any",
}

export interface FuncArgument {
  id: string;
  name: string;
  kind: FuncArgumentKind;
  elementKind?: FuncArgumentKind;
}

export interface FuncSummary {
  funcId: FuncId;
  kind: FuncKind;
  name: string;
  display_name: string | null;
  is_locked: boolean;
}

export interface FuncCode {
  funcId: FuncId;
  code: string;
  types: string;
}

export interface AttributeArgumentBinding {
  funcArgumentId: FuncArgumentId | null;
  attributePrototypeArgumentId: AttributePrototypeArgumentId | null;
  propId: PropId | null;
  inputSocketId: OutputSocketId | null;
}

export enum FuncBindingKind {
  Action = "Action",
  Attribute = "Attribute",
  Authentication = "Authentication",
  CodeGeneration = "CodeGeneration",
  Qualification = "Qualification",
}

export interface Action {
  bindingKind: FuncBindingKind.Action;
  // uneditable
  funcId: FuncId | null;
  schemaVariantId: SchemaVariantId | null;
  actionPrototypeId: ActionPrototypeId | null;
  // editable
  kind: ActionKind | null;
}

export interface Attribute {
  bindingKind: FuncBindingKind.Attribute;
  // uneditable
  funcId: FuncId | null;
  actionPrototypeId: ActionPrototypeId | null;
  // needed on create
  schemaVariantId: SchemaVariantId | null;
  componentId: ComponentId | null;
  // editable
  propId: PropId | null;
  outputSocketId: OutputSocketId | null;
  argumentBingindgs: AttributeArgumentBinding[];
}

export interface Authentication {
  bindingKind: FuncBindingKind.Authentication;
  funcId: FuncId;
  schemaVariantId: SchemaVariantId;
}

export interface CodeGeneration {
  bindingKind: FuncBindingKind.CodeGeneration;
  funcId: FuncId | null;
  schemaVariantId: SchemaVariantId | null;
  actionPrototypeId: ActionPrototypeId | null;
  componentId: ComponentId | null;
  // editable
  inputs: LeafInputLocation[];
}

export interface Qualification {
  bindingKind: FuncBindingKind.Qualification;
  funcId: FuncId | null;
  schemaVariantId: SchemaVariantId | null;
  actionPrototypeId: ActionPrototypeId | null;
  componentId: ComponentId | null;
  // editable
  inputs: LeafInputLocation[];
}

export type FuncBinding =
  | Action
  | Attribute
  | Authentication
  | CodeGeneration
  | Qualification;

export type LeafInputLocation =
  | "code"
  | "deletedAt"
  | "domain"
  | "resource"
  | "secrets";
